File: APPLICATION/services/index_manager.py
"""
index_manager.py — Gestor de índices de acceso directo (.idx y .holes).

CONCEPTO ACADÉMICO: Índice de Acceso Directo
═══════════════════════════════════════════════════════════════════════════════
El índice actúa como una tabla de mapeo que relaciona cada clave primaria
con su posición física (byte offset) dentro del archivo .dat:

    Archivo .idx en disco:
    ┌──────────────────────────────────────────────────────────────┐
    │  n_entradas (4 bytes)                                        │ ← Cabecera
    ├──────────────┬───────────────────────────────────────────────┤
    │  key=1 (4B)  │  offset=0   (8B)   → registro 1 en .dat[0]   │ ← Entrada 1
    │  key=2 (4B)  │  offset=97  (8B)   → registro 2 en .dat[97]  │ ← Entrada 2
    │  key=3 (4B)  │  offset=194 (8B)   → registro 3 en .dat[194] │ ← Entrada 3
    └──────────────┴───────────────────────────────────────────────┘

    En memoria: dict{ 1→0, 2→97, 3→194 }

BÚSQUEDA O(1):
    offset = index[id]          # lookup en dict
    file.seek(offset)           # posicionamiento directo
    data   = file.read(SIZE)    # lectura de registro de tamaño fijo

ARCHIVO .holes — Reutilización de Espacios Libres:
    Cuando se elimina un registro lógicamente, su offset queda disponible
    para ser reutilizado en la próxima inserción, evitando fragmentación:
    ┌───────────────────────┐
    │  n_holes (4 bytes)    │
    ├───────────────────────┤
    │  offset=97  (8 bytes) │  ← espacio libre del registro eliminado
    │  offset=291 (8 bytes) │
    └───────────────────────┘
"""
import struct
from pathlib import Path
from typing import Dict, List, Optional

# ── Formato de una entrada de índice: clave (4B) + offset (8B) = 12 bytes ────
IDX_ENTRY_FMT  = "!IQ"
IDX_ENTRY_SIZE = struct.calcsize(IDX_ENTRY_FMT)   # 12

# ── Formato de la cabecera: número de entradas (4B) ──────────────────────────
HDR_FMT  = "!I"
HDR_SIZE = struct.calcsize(HDR_FMT)               # 4


class IndexManager:
    """
    Gestiona el índice binario (.idx) y la lista de espacios libres (.holes)
    para un archivo de datos .dat.

    Provee:
      - Acceso O(1) a registros vía byte offset
      - Reutilización de slots de registros eliminados
      - Persistencia en disco en formato binario de longitud fija
    """

    def __init__(self, idx_path: Path, holes_path: Path) -> None:
        self._idx_path:   Path           = idx_path
        self._holes_path: Path           = holes_path
        self._index:      Dict[int, int] = {}    # { id → byte_offset }
        self._holes:      List[int]      = []    # [ byte_offset_libre, ... ]
        self._load_index()
        self._load_holes()

    # ════════════════════════════════════════════════════════════════════════
    # CARGA DESDE DISCO
    # ════════════════════════════════════════════════════════════════════════

    def _load_index(self) -> None:
        """
        Lee el archivo .idx y puebla el diccionario _index.

        Formato:
            [4 bytes: n_entradas] [ (4B key + 8B offset) × n_entradas ]
        """
        self._index.clear()
        if not self._idx_path.exists():
            return

        with open(self._idx_path, "rb") as f:
            hdr = f.read(HDR_SIZE)
            if len(hdr) < HDR_SIZE:
                return
            (n,) = struct.unpack(HDR_FMT, hdr)

            for _ in range(n):
                entry = f.read(IDX_ENTRY_SIZE)
                if len(entry) < IDX_ENTRY_SIZE:
                    break
                key, offset = struct.unpack(IDX_ENTRY_FMT, entry)
                self._index[key] = offset

    def _load_holes(self) -> None:
        """
        Lee el archivo .holes y puebla la lista _holes.

        Formato:
            [4 bytes: n_holes] [ 8B offset × n_holes ]
        """
        self._holes.clear()
        if not self._holes_path.exists():
            return

        with open(self._holes_path, "rb") as f:
            hdr = f.read(HDR_SIZE)
            if len(hdr) < HDR_SIZE:
                return
            (n,) = struct.unpack(HDR_FMT, hdr)

            for _ in range(n):
                raw = f.read(8)
                if len(raw) < 8:
                    break
                (offset,) = struct.unpack("!Q", raw)
                self._holes.append(offset)

    # ════════════════════════════════════════════════════════════════════════
    # PERSISTENCIA EN DISCO
    # ════════════════════════════════════════════════════════════════════════

    def _save_index(self) -> None:
        """Escribe el índice completo al archivo .idx."""
        self._idx_path.parent.mkdir(parents=True, exist_ok=True)
        with open(self._idx_path, "wb") as f:
            f.write(struct.pack(HDR_FMT, len(self._index)))
            # Ordenar por clave para facilitar búsqueda binaria futura
            for key in sorted(self._index):
                f.write(struct.pack(IDX_ENTRY_FMT, key, self._index[key]))

    def _save_holes(self) -> None:
        """Escribe la lista de holes al archivo .holes."""
        self._holes_path.parent.mkdir(parents=True, exist_ok=True)
        with open(self._holes_path, "wb") as f:
            f.write(struct.pack(HDR_FMT, len(self._holes)))
            for offset in self._holes:
                f.write(struct.pack("!Q", offset))

    # ════════════════════════════════════════════════════════════════════════
    # API PÚBLICA
    # ════════════════════════════════════════════════════════════════════════

    def get_offset(self, key: int) -> Optional[int]:
        """
        Retorna el byte offset del registro con la clave dada.

        Complejidad: O(1) — lookup en diccionario Python.

        Args:
            key: Clave primaria (ID del registro)

        Returns:
            Byte offset en el archivo .dat, o None si no existe
        """
        return self._index.get(key)

    def add(self, key: int, offset: int) -> None:
        """
        Registra key→offset en el índice y persiste en disco.

        Args:
            key:    Clave primaria
            offset: Byte offset en el archivo .dat
        """
        self._index[key] = offset
        self._save_index()

    def remove(self, key: int) -> Optional[int]:
        """
        Elimina la entrada del índice y agrega el offset a los holes.

        Al liberar el offset, queda disponible para ser reutilizado
        en la próxima inserción, evitando desperdiciar espacio en disco.

        Args:
            key: Clave primaria a eliminar

        Returns:
            Byte offset liberado, o None si no existía
        """
        offset = self._index.pop(key, None)
        if offset is not None:
            self._holes.append(offset)
            self._save_index()
            self._save_holes()
        return offset

    def get_free_slot(self) -> Optional[int]:
        """
        Obtiene y consume un slot libre (espacio reutilizable).

        Returns:
            Byte offset de un slot libre, o None si no hay ninguno
        """
        if self._holes:
            offset = self._holes.pop(0)
            self._save_holes()
            return offset
        return None

    def get_all(self) -> Dict[int, int]:
        """Retorna copia del índice completo { key → offset }."""
        return dict(self._index)

    def get_all_keys(self) -> List[int]:
        """Retorna lista de todas las claves activas."""
        return list(self._index.keys())

    def exists(self, key: int) -> bool:
        """Verifica si una clave está en el índice."""
        return key in self._index

    def count(self) -> int:
        """Número de entradas activas en el índice."""
        return len(self._index)

    def next_id(self) -> int:
        """
        Genera el próximo ID disponible (max_id_actual + 1).

        Returns:
            1 si no hay registros, max+1 en caso contrario
        """
        return max(self._index.keys(), default=0) + 1

    def reload(self) -> None:
        """Recarga índice y holes desde disco (útil post-compactación)."""
        self._load_index()
        self._load_holes()

    def debug_dump(self) -> str:
        """Representación textual del índice para depuración."""
        lines = [f"IndexManager({self._idx_path.name})"]
        lines.append(f"  Entradas activas: {len(self._index)}")
        lines.append(f"  Holes disponibles: {len(self._holes)}")
        for k, v in sorted(self._index.items()):
            lines.append(f"    key={k:6d} → offset={v:10d} bytes")
        return "\n".join(lines)

File: APPLICATION-JAVA/src/main/java/com/minimarket/services/IndexManager.java
package com.minimarket.services;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.logging.Logger;

/**
 * Gestor de índices de acceso directo (.idx) y espacios libres (.holes).
 *
 * CONCEPTO ACADEMICO: Indice de Acceso Directo
 * El indice mapea cada clave primaria a su byte offset en el archivo .dat,
 * permitiendo busqueda O(1) mediante RandomAccessFile.seek().
 *
 * Formato del archivo .idx:
 *   [4 bytes: n_entradas][12 bytes por entrada: 4B clave + 8B offset] x n
 *
 * Formato del archivo .holes:
 *   [4 bytes: n_holes][8 bytes por offset] x n
 *
 * Compatible byte a byte con el formato Python (big-endian, DataOutputStream
 * produce el mismo orden que struct.pack "!IQ").
 */
public class IndexManager {

    private static final Logger LOG = Logger.getLogger(IndexManager.class.getName());

    private final Path idxPath;
    private final Path holesPath;
    private final Map<Integer, Long> index; // clave → byte offset
    private final List<Long>         holes; // offsets libres reutilizables

    public IndexManager(Path idxPath, Path holesPath) {
        this.idxPath   = idxPath;
        this.holesPath = holesPath;
        this.index     = new TreeMap<>(); // TreeMap mantiene el orden por clave
        this.holes     = new ArrayList<>();
        loadIndex();
        loadHoles();
    }

    // ════════════════════════════════════════════════════════════════════════
    // CARGA DESDE DISCO
    // ════════════════════════════════════════════════════════════════════════

    private void loadIndex() {
        index.clear();
        if (!Files.exists(idxPath)) return;

        try (DataInputStream dis = new DataInputStream(
                new BufferedInputStream(Files.newInputStream(idxPath)))) {

            int n = dis.readInt(); // número de entradas
            for (int i = 0; i < n; i++) {
                int  key    = dis.readInt();
                long offset = dis.readLong();
                index.put(key, offset);
            }
        } catch (IOException e) {
            LOG.warning("No se pudo cargar el indice " + idxPath + ": " + e.getMessage());
        }
    }

    private void loadHoles() {
        holes.clear();
        if (!Files.exists(holesPath)) return;

        try (DataInputStream dis = new DataInputStream(
                new BufferedInputStream(Files.newInputStream(holesPath)))) {

            int n = dis.readInt();
            for (int i = 0; i < n; i++) {
                holes.add(dis.readLong());
            }
        } catch (IOException e) {
            LOG.warning("No se pudo cargar holes " + holesPath + ": " + e.getMessage());
        }
    }

    // ════════════════════════════════════════════════════════════════════════
    // PERSISTENCIA EN DISCO
    // ════════════════════════════════════════════════════════════════════════

    private void saveIndex() {
        ensureParent(idxPath);
        try (DataOutputStream dos = new DataOutputStream(
                new BufferedOutputStream(Files.newOutputStream(idxPath)))) {

            dos.writeInt(index.size());
            for (Map.Entry<Integer, Long> e : index.entrySet()) {
                dos.writeInt(e.getKey());
                dos.writeLong(e.getValue());
            }
        } catch (IOException e) {
            LOG.severe("No se pudo guardar el indice " + idxPath + ": " + e.getMessage());
        }
    }

    private void saveHoles() {
        ensureParent(holesPath);
        try (DataOutputStream dos = new DataOutputStream(
                new BufferedOutputStream(Files.newOutputStream(holesPath)))) {

            dos.writeInt(holes.size());
            for (long offset : holes) {
                dos.writeLong(offset);
            }
        } catch (IOException e) {
            LOG.severe("No se pudo guardar holes " + holesPath + ": " + e.getMessage());
        }
    }

    // ════════════════════════════════════════════════════════════════════════
    // API PUBLICA
    // ════════════════════════════════════════════════════════════════════════

    /** Retorna el byte offset del registro con la clave dada, o -1 si no existe. */
    public long getOffset(int key) {
        Long off = index.get(key);
        return off != null ? off : -1L;
    }

    /** Registra clave → offset en el indice y persiste. */
    public void add(int key, long offset) {
        index.put(key, offset);
        saveIndex();
    }

    /**
     * Elimina la entrada del indice y agrega el offset a los holes.
     * Retorna el offset liberado, o -1 si no existia.
     */
    public long remove(int key) {
        Long offset = index.remove(key);
        if (offset != null) {
            holes.add(offset);
            saveIndex();
            saveHoles();
            return offset;
        }
        return -1L;
    }

    /**
     * Obtiene y consume el primer slot libre disponible.
     * Retorna -1 si no hay ninguno (señal de append al final).
     */
    public long getFreeSlot() {
        if (holes.isEmpty()) return -1L;
        long offset = holes.remove(0);
        saveHoles();
        return offset;
    }

    /** Genera el proximo ID disponible (max + 1). */
    public int nextId() {
        if (index.isEmpty()) return 1;
        return Collections.max(index.keySet()) + 1;
    }

    /** Retorna copia del mapa indice completo. */
    public Map<Integer, Long> getAll() { return new TreeMap<>(index); }

    /** Verifica si una clave esta en el indice. */
    public boolean exists(int key) { return index.containsKey(key); }

    /** Numero de entradas activas. */
    public int count() { return index.size(); }

    /** Recarga indice y holes desde disco (util tras compactacion). */
    public void reload() {
        loadIndex();
        loadHoles();
    }

    // Acceso interno para compactacion (solo FileManager lo usa)
    Map<Integer, Long>  getIndexMap()  { return index; }
    List<Long>          getHolesList() { return holes; }
    void                persistAll()   { saveIndex(); saveHoles(); }

    // ── Utilidades ───────────────────────────────────────────────────────────

    private static void ensureParent(Path path) {
        try {
            Files.createDirectories(path.getParent());
        } catch (IOException ignored) { }
    }

    public String debugDump() {
        StringBuilder sb = new StringBuilder();
        sb.append("IndexManager(").append(idxPath.getFileName()).append(")\n");
        sb.append("  Entradas activas: ").append(index.size()).append("\n");
        sb.append("  Holes disponibles: ").append(holes.size()).append("\n");
        for (Map.Entry<Integer, Long> e : index.entrySet()) {
            sb.append(String.format("    key=%6d → offset=%10d bytes%n", e.getKey(), e.getValue()));
        }
        return sb.toString();
    }
}
